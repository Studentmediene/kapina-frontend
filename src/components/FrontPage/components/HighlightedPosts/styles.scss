@import 'variables.scss';

.container {
  display: flex;
  flex-direction: column;
  align-items: flex-start;

  // Negative margin to force element to take
  // full width of container on smaller devices.
  margin: 0 -2.6%;
  padding: 20px 0 30px 0;

  color: $color-text-light;
  background-color: $color-radiorevolt-lightgrey;
}

.title {
  font-size: $frontpage-title-size;
  margin: 0 0 10px 2.5%;
}

.imageText {
  position: absolute;
  text-decoration: none;
  color: $color-text-light;
  background-color: rgba(0, 0, 0, 0.77);
  font-size: 1.1em;
  bottom: 0;
  left: 0;
  padding: 5px;
  width: 100%;
}

.postContainer {
  display: flex;
  flex-direction: row;

  overflow-x: scroll;
  scrollbar-width: none;

  &::-webkit-scrollbar {
    display: none;
  }
}

.endPadding {
  // Padding fix on end of scrollable content, to force
  // padding within scrollable box
  width: 2.5%;
  padding-right: 2.5%;
}

.highlightedPost {
  display: inline-block;
  position: relative;
  margin: 0 0 0 2.5%;
  width: 70%;
  flex-shrink: 0;
}

@media screen and (min-width: $breakpoint-large) {
  .container {
    margin: 0 auto;
    max-width: $breakpoint-large;
    grid-column-start: 1;
    grid-column-end: 3;
  }

  .imageText {
    font-size: 1.5em;
  }
}
