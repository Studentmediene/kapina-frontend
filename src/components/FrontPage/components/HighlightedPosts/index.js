import React from 'react';
import PropTypes from 'prop-types';

import styles from './styles.scss';
import ImageLink from 'components/common/ImageLink';
import transparentPixel from './assets/transparent-pixel.png';

class HighlightedPosts extends React.Component {
  constructor(props) {
    super(props);

    this.drag = this.drag.bind(this);
    this.dragStart = this.dragStart.bind(this);

    // fields for keeping state while dragging
    this.postContainerRef = React.createRef();
    const dragGhostImage = new Image();
    dragGhostImage.src = transparentPixel;
    this.dragGhostImage = dragGhostImage;
    this.initialPointerX = null;
    this.initialContainerScroll = null;
  }

  drag(event) {
    if (event.pageX === 0) {
      return;
    }
    const container = this.postContainerRef.current;
    const pointerMovement = event.pageX - this.initialPointerX;

    container.scrollTo({
      left: this.initialContainerScroll - pointerMovement,
    });
  }

  dragStart(event) {
    // Set the "ghost image" shown on drag to a transparent pixel
    event.nativeEvent.dataTransfer.setDragImage(this.dragGhostImage, 0, 0);

    // Save the initial position of pointer and container on drag start
    this.initialPointerX = event.pageX;
    this.initialContainerScroll = this.postContainerRef.current.scrollLeft;
  }
  render() {
    if (this.props.posts && this.props.posts.length > 0) {
      const posts = this.props.posts.map(post => {
        return (
          <div
            key={post.title}
            className={styles.highlightedPost}
            key={post.id}
          >
            <ImageLink
              images={post.croppedImages}
              link={`/post/${post.slug}`}
              imageDescription={post.title}
            >
              <div className={styles.imageText}>{post.title}</div>
            </ImageLink>
          </div>
        );
      });
      // Add extra padding to end of array
      posts.push(<span className={styles.endPadding} key={'padding'} />);
      return (
        <div className={styles.container}>
          <h2 className={styles.title}>Anbefalte saker</h2>
          <div
            className={styles.postContainer}
            ref={this.postContainerRef}
            onDrag={this.drag}
            onDragStart={this.dragStart}
          >
            {posts}
          </div>
        </div>
      );
    } else {
      return null;
    }
  }
}

HighlightedPosts.propTypes = {
  posts: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.arrayOf(
      PropTypes.shape({
        croppedImages: PropTypes.shape({
          small: PropTypes.string.isRequired,
          medium: PropTypes.string.isRequired,
          large: PropTypes.string.isRequired,
        }).isRequired,
        slug: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired,
      }),
    ),
  ]).isRequired,
};

export default HighlightedPosts;
