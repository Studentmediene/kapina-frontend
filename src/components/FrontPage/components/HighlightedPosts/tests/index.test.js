import React from 'react';
import { shallow } from 'enzyme';

import HighlightedPosts from '../';

describe('<HighlightedPosts />', () => {
  const emptyProps = {
    posts: false,
  };

  it('renders correctly when empty', () => {
    const tree = shallow(<HighlightedPosts {...emptyProps} />);
    expect(tree).toMatchSnapshot();
  });

  const noPostsProps = {
    posts: [],
  };

  it('renders correctly when there are no posts', () => {
    const tree = shallow(<HighlightedPosts {...noPostsProps} />);
    expect(tree).toMatchSnapshot();
  });

  const props = {
    posts: [
      {
        id: 364,
        croppedImages: {
          large: '/large1.jpg',
          medium: '/medium1.jpg',
          small: '/small1.jpg',
        },
        title: 'title1',
        slug: 'slug1',
      },
      {
        id: 362,
        croppedImages: {
          large: '/large2.jpg',
          medium: '/medium2.jpg',
          small: '/small2.jpg',
        },
        title: 'title2',
        slug: 'slug2',
      },
    ],
  };

  it('renders correctly', () => {
    const tree = shallow(<HighlightedPosts {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
