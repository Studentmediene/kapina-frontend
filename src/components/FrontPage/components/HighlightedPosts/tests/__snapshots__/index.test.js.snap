// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HighlightedPosts /> renders correctly 1`] = `
<div
  className="container"
>
  <h2
    className="title"
  >
    Vi anbefaler
  </h2>
  <div
    className="postContainer"
    onDrag={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="highlightedPost"
      key="364"
    >
      <ImageLink
        imageDescription="title1"
        images={
          Object {
            "large": "/large1.jpg",
            "medium": "/medium1.jpg",
            "small": "/small1.jpg",
          }
        }
        link="/post/slug1"
      >
        <div
          className="imageText"
        >
          title1
        </div>
      </ImageLink>
    </div>
    <div
      className="highlightedPost"
      key="362"
    >
      <ImageLink
        imageDescription="title2"
        images={
          Object {
            "large": "/large2.jpg",
            "medium": "/medium2.jpg",
            "small": "/small2.jpg",
          }
        }
        link="/post/slug2"
      >
        <div
          className="imageText"
        >
          title2
        </div>
      </ImageLink>
    </div>
    <span
      className="endPadding"
      key="padding"
    />
  </div>
</div>
`;

exports[`<HighlightedPosts /> renders correctly when empty 1`] = `""`;

exports[`<HighlightedPosts /> renders correctly when there are no posts 1`] = `""`;
